### PR TITLE
Fix #329 - Support `ALTER` queries of `PARTITIONS`

### LINKED ISSUE #329: SQL parser: "ALTER TABLE ADD/DROP/REORGANIZE PARTITION" error
### Describe the bug

According to [MySQL Reference Manual - 13.1.8.1 ALTER TABLE Partition Operations](https://dev.mysql.com/doc/refman/5.7/en/alter-table-partition-operations.html),we can use 
```SQL
ALTER TABLE t1 ADD PARTITION (PARTITION p3 VALUES LESS THAN (2002));
ALTER TABLE t1 DROP PARTITION p0, p1;
```
and so on to change partition.
**But when I copy these statements to the SQL console (including the `console` in the bottom  and `SQL` on the top),it reminds me:**
```
1.This option conflicts with "ADD". (near "PARTITION" at position 19)
2.Missing comma before start of a new alter operation. (near "PARTITION" at position 30)
```

But when I go these statements,it worked(I have create table like the example in the MySQL document).

**It seems that the SQL parser in phpMyAdmin doesn't parse these statements properly.**

### To Reproduce

1. Create a table in a database like what are in MySQL Document (link is above)
```SQL
CREATE TABLE t1 (
    id INT,
    year_col INT
)
PARTITION BY RANGE (year_col) (
    PARTITION p0 VALUES LESS THAN (1991),
    PARTITION p1 VALUES LESS THAN (1995),
    PARTITION p2 VALUES LESS THAN (1999)
);
```
2. Input these statements in console:
```SQL
ALTER TABLE t1 ADD PARTITION (PARTITION p3 VALUES LESS THAN (2002));
ALTER TABLE t1 DROP PARTITION p0, p1;
```
3. You will find it underlined some words (like "partition").
4. Press Ctrl+Enter to excute,and you will find it queried ok.

### Expected behavior

phpMyAdmin parse the ALTER TABLE Partition Operations in MySQL Document properly.

### Screenshots

ScreenShot1

### Server configuration

- Operating system: Ubuntu Server 20.04
- Web server: Nginx 1.18.0
- Database version: Mariadb 10.3.25
- PHP version: 7.4.3
- phpMyAdmin version: 5.0.4

### Client configuration

- Browser: Chrome 81
- Operating system: Windows 7

Thanks.

### PR DESCRIPTION
**Pull request by iifawzi**
Hi, this should fix the issue https://github.com/phpmyadmin/sql-parser/issues/329, as per https://github.com/phpmyadmin/sql-parser/pull/376. 

Signed-off-by: Fawzi E. Abdulfattah <iifawzie@gmail.com>

### CODE REVIEW

**Comment by iifawzi** [comment]
I thought that the baseline issue could be fixed by merging from the master branch, but it didn't change anything :)

**Review by williamdes — APPROVED** [review]
This looks good, thank you !

File: src/Components/AlterOperation.php [hunk 1]
```diff
@@ -203,15 +211,18 @@ class AlterOperation extends Component
     public $unknown = [];
 
     /**
-     * @param OptionsArray $options options of alter operation
-     * @param Expression   $field   altered field
-     * @param Token[]      $unknown unparsed tokens found at the end of operation
+     * @param OptionsArray              $options    options of alter operation
+     * @param Expression|string|null    $field      altered field
+     * @param Token[]                   $unknown    unparsed tokens found at the end of operation
+     * @param Component[]|ArrayObj|null $partitions partitions definition found in the operation
```
Comment (williamdes) [inline src/Components/AlterOperation.php:217]: The order is wrong

**Review by williamdes — COMMENTED** [review]
Could you rebase on master branch ?
I made a baseline update

**Comment by iifawzi** [comment]
Rebased, but still the same error, this drives me crazy.

**Comment by williamdes** [comment]
> Rebased, but still the same error, this drives me crazy.

I did re do the update with `php7.4 ./vendor/bin/phpstan -b`

File: src/Components/AlterOperation.php [hunk 2]
```diff
@@ -362,6 +385,46 @@ public static function parse(Parser $parser, TokensList $list, array $options =
...
+                    if (empty($ret->field)) {
+                        $ret->field = '';
+                    }
+
+                    $ret->field .= $token->type === Token::TYPE_WHITESPACE ? ' ' : $token->token;
+                } elseif ($partitionState === 2) {
+                    $ret->partitions = ArrayObj::parse(
+                        $parser,
+                        $list,
+                        ['type' => 'PhpMyAdmin\\SqlParser\\Components\\PartitionDefinition']
```
Comment (williamdes) [inline src/Components/AlterOperation.php:425]:
```suggestion
                        ['type' => PartitionDefinition::class]
```
Should work ?

Comment (iifawzi) [reply src/Components/AlterOperation.php]: seems like legitimate, was used this way before, would this help fix the issue?

Comment (iifawzi) [reply src/Components/AlterOperation.php]: it's now showing just a single error, progresses :'d

Comment (williamdes) [reply src/Components/AlterOperation.php]: > seems legitimate, was used this way before, would this help fix the issue?

Yeah, its an old PHP compat issue, probably from QA

**Review by williamdes — APPROVED** [review]
Looks ready for a merge, thank you !

**Pull request merged**